### PR TITLE
restore: add fileregex parameter

### PR DESCRIPTION
**Pull request by SamuelBoerlin**
### Thank you for contributing to the Bareos Project!

I'm currently using python-bareos to run automated restores and I'd like to filter by regex which files are restored. As far as I can tell this is currently not possible to do via `restore` command (except when records were pruned), hence this PR.

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)
- [Boy Scout Rule](https://docs.bareos.org/DeveloperGuide/generaldevel.html#boy-scout-rule)

### Checklist for the _reviewer_ of the PR (will be processed by the Bareos team)
Make sure you check/merge the PR using `devtools/pr-tool` to have some simple automated checks run and a proper changelog record added.

##### General
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
~~Check backport line~~
- [x] Required backport PRs have been created

##### Source code quality
- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR

##### Tests
- [x] Decision taken that a test is required (if not, then remove this paragraph)
- [x] The choice of the type of test (unit test or systemtest) is reasonable
- [x] Testname matches exactly what is being tested
- [x] On a fail, output of the test leads quickly to the origin of the fault

### CODE REVIEW

**Review by sebsura — CHANGES_REQUESTED** [review]
This is a great change.  You need to fix your commit formatting though.  `pr-tool check` reports the following problems:
```
4af68f0e0 systemtests:add RestoreDirectory variable to allow changing default restore directory: headline too long
b8c4d33d9 docs: add fileregex parameter to restore command documentation: headline too long
```

**Comment by SamuelBoerlin** [comment]
Is it okay now?  
I tried running `pr-tool check` to confirm but it fails with `Unknown JSON field: "headRefOid"`.

**Comment by sebsura** [comment]
> Is it okay now? I tried running `pr-tool check` to confirm but it fails with `Unknown JSON field: "headRefOid"`.

Did you run `pipenv sync` ? Are you working in a full bareos clone or did you do a shallow clone / single branch clone ?

**Comment by SamuelBoerlin** [comment]
> > Is it okay now? I tried running `pr-tool check` to confirm but it fails with `Unknown JSON field: "headRefOid"`.
> 
> Did you run `pipenv sync` ? if not you can find instructions on how to setup pr-tool inside `devtools/pip-tools/README.md`. Are you working in a full bareos clone or did you do a shallow clone / single branch clone ?

No, I did not run `pipenv sync`, maybe that's it. Thanks!

**Comment by sebsura** [comment]
Sorry for the late reply.  Yes it is still possible to get it into 23.  While we plan to release a preview of 23.0.0 very soon, it will still take some time for the real release.

While testing the new feature a bit I noticed the following problem:
The file tree is still getting build.  This can take some time and selecting items in that tree can be very awkward since you may select items that do not match the supplied regex.  We would like it to either behave similarly to `file=` or `dir=` in that it does not build the tree and requires no manual selection or instead the file tree needs to not insert items not matching the regex.

**Comment by SamuelBoerlin** [comment]
> Sorry for the late reply. Yes it is still possible to get it into 23. While we plan to release a preview of 23.0.0 very soon, it will still take some time for the real release.
> 
> While testing the new feature a bit I noticed the following problem: The file tree is still getting build. This can take some time and selecting items in that tree can be very awkward since you may select items that do not match the supplied regex. We would like it to either behave similarly to `file=` or `dir=` in that it does not build the tree and requires no manual selection or instead the file tree needs to not insert items not matching the regex.

That's great, thanks.
I hadn't realized this behaviour and I totally agree with your suggestion. I think it'd make most sense to make it behave like `file=` or `dir=` where it doesn't ask the user to select files nor build the file tree. I'll try to implement that.

**Comment by SamuelBoerlin** [comment]
I've now changed fileregex to behave more like the file parameter. If no JobIds are specified then it selects the most recent JobIds (before specified date) to restore from without asking the user or building the file tree.

File: systemtests/tests/restore/testrunner-restore-fileregex [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+# Use own restore directory to ensure it is always
+# clean before restore
+RestoreDirectory="${tmp}/fileregex-restores"
+
+start_test
+
+fileregex=".*/ot[h]er-.*"
+
+matched_files=$(find "$BackupDirectory/" -type f -regex "$fileregex")
```
Comment (sebsura) [inline systemtests/tests/restore/testrunner-restore-fileregex:27]: This does not work if paths contain whitespaces!
An alternative could be
```
mapfile -d $'\0' matched_files < <(find "$BackupDirectory/" -type f \
					-regex "$fileregex" -print0)
```

File: systemtests/tests/restore/testrunner-restore-fileregex [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+if [ -z "$matched_files" ]; then
+  echo "FileRegex did not match any files in test data"
+  exit 1
+fi
```
Comment (sebsura) [inline systemtests/tests/restore/testrunner-restore-fileregex:32]: For the array version:
```suggestion
if [ "${#matched_files[@]}" -eq 0 ]; then
  echo "FileRegex did not match any files in test data"
  exit 1
fi
```

**Review by sebsura — CHANGES_REQUESTED** [review]
The code changes are great.  I added some comments to the doc & test.  Once those are addressed, it will be approved.

File: systemtests/tests/restore/testrunner-restore-fileregex [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+# Use own restore directory to ensure it is always
+# clean before restore
+RestoreDirectory="${tmp}/fileregex-restores"
+
+start_test
+
+fileregex=".*/ot[h]er-.*"
+
+matched_files=$(find "$BackupDirectory/" -type f -regex "$fileregex")
+
```
[the remark this answers is not in the document]
Comment (SamuelBoerlin) [reply systemtests/tests/restore/testrunner-restore-fileregex:29]: Done

File: systemtests/tests/restore/testrunner-restore-fileregex (hunk 1 above)
Comment (SamuelBoerlin) [reply systemtests/tests/restore/testrunner-restore-fileregex]: Done

**Comment by SamuelBoerlin** [comment]
Thanks for your suggestions. I've changed the file regex to something else so that it also contains files with special characters and whitespaces. There was also a small bug in `check_restore_files_diff()` where it would interpret the file names as regex pattern, causing it to fail e.g. for files with an unescaped backslash in it.

**Review by sebsura — APPROVED** [review]

**Pull request merged**